### DISCUSSION TITLE
webpack 5 web worker support for CORS?

### DISCUSSION THREAD

**Discussion by piotr-oles** [webpack/webpack#14648] [Q&A] [answered]
Hi! I would like to use the new worker import syntax:
```typescript
const worker = new Worker(new URL('./worker', import.meta.url));
``` 
to share code between the main thread and the worker. 

Unfortunately, the project I'm working on uses CDN on a separate domain, and I get CORS error. Would it be possible to add some option in webpack configuration or somewhere else to use work-around with `importScript`?
The transformation could be:
### From:
```typescript
const worker = new Worker(new URL('./worker', import.meta.url));
```

### To:
```typescript
const objectURL = URL.createObjectURL(
    new Blob([`importScripts(${JSON.stringify(generatedUrlToWorker)});`], { type: 'application/javascript' })
);
const worker = new Worker(objectURL);
URL.revokeObjectURL(objectURL);
```

**Comment by alexander-akait** [comment]
https://github.com/webpack/webpack/discussions/14198#discussioncomment-1514273

**Reply by piotr-oles** [reply]
Will this work-around share common code as the default syntax?

**Reply by alexander-akait** [reply]
What do you mean? It is not workaround, using `importScripts` has a lot of limitations

**Reply by piotr-oles** [reply]
Agree, that `importScripts` has a lot of limitations, and I'm open for other solutions. I just want to share common code as described in https://github.com/webpack/webpack/issues/6472 . I'm not sure if it will work with the `new URL()` without `Worker` constructor wrapping it, that's why I ask :)

**Reply by piotr-oles** [reply]
Also, I was trying the solution with `asset/resource` and it seems like webpack doesn't process it as a JavaScript file. I have unresolved imports in the response:
![image](https://user-images.githubusercontent.com/8374473/140326082-ab928263-c74a-4df0-aa88-4da4ed5150e6.png)

**Reply by alexander-akait** [reply]
Unfortunately you need to use `new URL()` inside `new Worker()`, otherwise it is not statically analyzable

**Reply by alexander-akait** [reply]
> Also, I was trying the solution with asset/resource and it seems like webpack doesn't process it as a JavaScript file. I have unresolved imports in the response:

It means you have something wrong in your configuration, double check it

**Reply by LeeKevin** [reply]
> Also, I was trying the solution with asset/resource and it seems like webpack doesn't process it as a JavaScript file. I have unresolved imports in the response:

@alexander-akait I have the same problem. Imports are not resolved with `asset/resource` even when loaders are applied (like babel loader)

**Comment by piotr-oles** [comment, marked as answer]
I think I found a nice work-around. I created custom worker class in a separate file:
```typescript
export class CorsWorker {
    private readonly worker: Worker;

    constructor(url: string | URL) {
        const objectURL = URL.createObjectURL(
            new Blob([`importScripts(${JSON.stringify(url.toString())});`], {
                type: 'application/javascript'
            })
        );
        this.worker = new Worker(objectURL);
        URL.revokeObjectURL(objectURL);
    }

    getWorker() {
        return this.worker;
    }
}
```
and then I imported it in another file as Worker:

```typescript
import { CorsWorker as Worker } from './cors-worker';

// aliasing CorsWorker to Worker makes it statically analyzable
const corsWorker = new Worker(new URL('../worker', import.meta.url));
// tada!
const worker = corsWorker.getWorker();
```

**Reply by alexander-akait** [reply]
Yes, good solution

**Reply by piotr-oles** [reply]
@alexander-akait thanks for pointing me to this direction :)

**Reply by piotr-oles** [reply]
@alexander-akait I found an issue with this solution - when I have some shared chunks between main thread and worker, and a relative publicPath (without domain name), it uses document domain, instead of script domain (CDN).

To better illustrate this problem, lets assume that we have 2 origins:
 1. document origin: https://company.com/
 2. cdn origin: https://cdn.company.com/
 
 All JavaScript files are served from the CDN origin. When we construct new Worker using Blob, it inherits document origin. It means that `importScripts('/my-script.js')` will resolve to `importScripts('https://company.com/my-script.js')` instead of `importScripts('https://cdn.company.com/my-script.js')`
 
I found this runtime module responsible for setting public path. 
```javascript
    /******/ 	/* webpack/runtime/publicPath */
    /******/ 	(() => {
    /******/ 		__webpack_require__.p = "/static/";
    /******/ 	})();
    /******/
```
I tired set `__webpack_public_path__` as webpack docs suggests, but it doesn't change anything (and I would be surprised if it would, because there is no reference to `__webpack_public_path__` in the worker code).
Do you have any ideas how I could set custom publicPath in the worker scope?

**Reply by alexander-akait** [reply]
You can move it to own entry and set publicPath for entry

**Reply by piotr-oles** [reply]
I would like to avoid this - it requires setting new entry for each worker. Also, I would like to preserve "relative" publicPath for worker, because we serve them from different cdn domains (for different environments and also versions preview). In this case, this would not solve the issue. Could we somehow overwrite this runtime chunk for workers to generate something like `__webpack_require__.p = __worker_origin__ + "/static/"`? I could set __worker_origin__ variable before importScripts call.

**Reply by alexander-akait** [reply]
You can rewrite public path on fly https://webpack.js.org/guides/public-path/#on-the-fly

**Reply by piotr-oles** [reply]
I was trying that, but it didn't work for me. The worker module doesn't have any reference to the `__webpack_public_path__` variable :/

**Reply by alexander-akait** [reply]
It means, it is impossible, move it to own entry, it will be valid solution

**Reply by piotr-oles** [reply]
It will not share common code as it does currently, so it's not good enough for me :/

**Reply by piotr-oles** [reply]
@sokra , do you have any ideas?

**Reply by ArturWierzbicki** [reply]
@piotr-oles I am struggling with the same issue - submitted a [PR](https://github.com/webpack/webpack/pull/14667) that would allow us to do something like below if used together with `publicPath: auto`

```ts
declare let __webpack_public_path__: string;
export class CorsWorker {
  private readonly worker: Worker;

  constructor(url: string | URL) {
    const objectURL = URL.createObjectURL(
      new Blob(
        [`__webpack_override_auto_public_path__ = '${__webpack_public_path__}'; importScripts('${url.toString()}');`],
        {
          type: 'application/javascript',
        }
      )
    );
    this.worker = new Worker(objectURL);
    URL.revokeObjectURL(objectURL);
  }

  getWorker() {
    return this.worker;
  }
}
```

**Reply by piotr-oles** [reply]
Thanks for that PR! I think that we shouldn't overwrite the full public path, but the worker origin.
You can for example have `publicPath` set to `/static/` - in this case `/static/` is a valid public path, but the origin is wrong. So I think we should detect if the URL is relative or absolute and in case of relative URLs transform it to absolute URLs using provided origin.
I will probably have some working code tomorrow, so I will share it :)

**Reply by piotr-oles** [reply]
I have a solution that works without modifying webpack source code:
```javascript
// src/webpack/plugins/cors-worker-plugin/CorsWorkerPlugin.js
const { RuntimeGlobals } = require('webpack');
const CorsWorkerPublicPathRuntimeModule = require('./CorsWorkerPublicPathRuntimeModule');

class CorsWorkerPlugin {
    /**
     * @param {import('webpack').Compiler} compiler
     */
    apply(compiler) {
        compiler.hooks.compilation.tap(
            'CorsWorkerPlugin',
            /**
             * @param {import('webpack').Compilation} compilation
             */
            compilation => {
                const getChunkLoading = chunk => {
                    const entryOptions = chunk.getEntryOptions();
                    return entryOptions &&
                        entryOptions.chunkLoading !== undefined
                        ? entryOptions.chunkLoading
                        : compilation.outputOptions.chunkLoading;
                };
                const getChunkPublicPath = chunk => {
                    const entryOptions = chunk.getEntryOptions();
                    return entryOptions && entryOptions.publicPath !== undefined
                        ? entryOptions.publicPath
                        : compilation.outputOptions.publicPath;
                };

                compilation.hooks.runtimeRequirementInTree
                    .for(RuntimeGlobals.publicPath)
                    .tap('CorsWorkerPlugin', chunk => {
                        if (getChunkLoading(chunk) === 'import-scripts') {
                            const publicPath = getChunkPublicPath(chunk);

                            if (publicPath !== 'auto') {
                                const module = new CorsWorkerPublicPathRuntimeModule(
                                    publicPath
                                );
                                compilation.addRuntimeModule(chunk, module);
                                return true;
                            }
                        }
                    });
            }
        );
    }
}

module.exports = CorsWorkerPlugin;
```
```javascript
// src/webpack/plugins/cors-worker-plugin/CorsWorkerPublicPathRuntimeModule.js
const { RuntimeGlobals, RuntimeModule } = require('webpack');

class CorsWorkerPublicPathRuntimeModule extends RuntimeModule {
    constructor(publicPath) {
        super('publicPath', RuntimeModule.STAGE_BASIC);
        this.publicPath = publicPath;
    }

    /**
     * @returns {string} runtime code
     */
    generate() {
        const { compilation, publicPath } = this;
        const isAbsolutePublicPath = (publicPath || '').indexOf('://') > 0;

        return `${RuntimeGlobals.publicPath} = ${
            isAbsolutePublicPath ? '' : '(__webpack_worker_origin__ || "") + '
        }${JSON.stringify(
            compilation.getPath(publicPath || '', {
                hash: compilation.hash || 'XXXX'
            })
        )};`;
    }
}

module.exports = CorsWorkerPublicPathRuntimeModule;
```
then in webpack.config.js:
```javascript
const CorsWorkerPlugin = require('./src/webpack/plugins/cors-webpack-plugin/CorsWebpackPlugin');
module.exports = {
  // ...
  plugins: [
    // ...
    new CorsWorkerPlugin()
  ]
}
```
and in the source code:
```typescript
// src/cors-worker.ts
export class Worker extends window.Worker {
    constructor(url: string | URL, options?: WorkerOptions) {
        const parsedUrl = new URL(url);
        const webpackWorkerOrigin = `__webpack_worker_origin__ = ${JSON.stringify(
            parsedUrl.origin
        )};`;
        const importScripts = `importScripts(${JSON.stringify(
            url.toString()
        )});`;
        const objectURL = URL.createObjectURL(
            new Blob([`${webpackWorkerOrigin};\n${importScripts}`], {
                type: 'application/javascript'
            })
        );
        super(objectURL, options);
        URL.revokeObjectURL(objectURL);
    }
}
```
Would be nice to integrate it into webpack, because it's a common use-case and should not require so much patching :)

**Reply by piotr-oles** [reply]
And it's still missing support for "auto" publicPath, but adding `CorsWorkerAutoPublicPathRuntimeModule` should be pretty straightforward

**Reply by CraigglesO** [reply]
LETS GOOOO DUDE. @piotr-oles this is insanely cool.

I have little experience working with webpack, can I make the plugin as a module rule rather than a plugin? Something like:

```js
module.exports = {
...
module: {
    rules: [
      {
              test: /\.worker\.js$/,
              loader: require.resolve('../plugins/cors-webpack-plugin/CorsWebpackPlugin.js')
      },

....

   ]
}
}
```

If I have any other questions can I keep posting them here?

Thanks,
Craig.

edit - nevermind, i get it more now. It works best as a plugin.

**Reply by ArturWierzbicki** [reply]
This is great, thanks @piotr-oles. Doesn't this:

```ts
  const webpackWorkerOrigin = `__webpack_worker_origin__ = ${JSON.stringify(
            parsedUrl.origin
        )};`;
```

 miss a case where we have a pathname that is not included in a relative `publicPath`? e.g. 

```
  pathname: '/cdn'
  origin: 'http://host.com'
  publicPath: 'scripts/'
```

exemplary script url: `http://host.com/cdn/scripts/chunk-123.js`

would you mind creating a repo with this plugin so we can easily attribute (and contribute to!) your work? :)

**Reply by piotr-oles** [reply]
I created a PR to add native support for cross-origin workers in webpack: https://github.com/webpack/webpack/pull/14680/files

**Reply by CraigglesO** [reply]
@piotr-oles - I'm now using your latest implementation above for my project. Went from 600kB+ gzip single file to less than 100kB+ total in chunks and the main thread is only 10kB. Thank you so much for this. Hopefully your PR goes through soon so I can switch to that.

**Reply by LeeKevin** [reply]
@piotr-oles I'm getting "The script at 'blob:...' failed to load". 

My worker loads code from another chunk, and that's not resolving nicely.

**Reply by piotr-oles** [reply]
Try to debug in Firefox - for some reason worker errors in Chrome are not very helpful

**Reply by OmarCastro** [reply]
Hello, I have been looking for a similar solution, but I had url origin problems too. After looking a this thread I got an idea and did test it.
I found that the importScripts urls are relative to the worker file, so what I had to do do is to override the importScript to use the worker file url as the base url. Here is the code I am using to solve it.

```javascript
export class CorsWorker {
  constructor(url) {
    const absoluteUrl = new URL(url, window.location.href).toString()
    const workerSource = `\
  const urlString = ${JSON.stringify(absoluteUrl)}
  const originURL = new URL(urlString)
  const originalImportScripts = self.importScripts
  self.importScripts = (url) => originalImportScripts.call(self, new URL(url, originURL).toString())
  importScripts(urlString);
`
    const blob = new Blob([workerSource], { type: 'application/javascript' })
    const objectURL = URL.createObjectURL(blob)
    this.worker = new Worker(objectURL)
    URL.revokeObjectURL(objectURL)
  }

  getWorker() {
    return this.worker
  }
}
```

The rest is to use the same way @piotr-oles described. I had no problems with relative paths with this solution

**Reply by Ketec** [reply]
End up with the same problem. It seems it's trying to load it from the local filesystem instead, not correctly parsing the URL - and standard import leads to cors errors due to different domain/port.

`Not allowed to load local resource: file:///home/user/projects/TestRemoteApp/src/app/test.worker`

**Reply by keshav-kishor** [reply]
@piotr-oles I tried your solution but I get this error:
`Uncaught SyntaxError: Cannot use 'import.meta' outside a module `
any idea?

**Reply by rahilvijay** [reply]
I am also getting a similar error about the import statement any solution for the same.

**Reply by piotr-oles** [reply]
Sorry, I don't know why you have such an error :/

**Reply by onyzhin** [reply]
I have tried to integrate the CorsWorker solution, but it does not work when importing a 3rd party library inside a web worker:

`"Failed to execute 'importScripts' on 'WorkerGlobalScope': The script at 'http://localhost:4201/worker.js' failed to load."`

It seems the browser throws an error if there is an import statement inside any of the scripts where the user is using importScripts().
Do you have any suggestions on how to handle this?

**Reply by Ketec** [reply]
If I recall, the new Module workers don't support importScripts anymore - only classic.
It now uses ES5 module imports.

**Reply by brotheroftux** [reply]
I think I found the ugliest workaround for dynamic public path issue. @OmarCastro 's solution didn't work for me, since webpack was emitting a chunk that would depend on other vendor chunks.

```typescript
export class CorsWorker {
    private readonly _worker: Worker;

    constructor(url: URL) {
        // @ts-expect-error
        const publicPath: string = __webpack_public_path__;

        const objectURL = URL.createObjectURL(
            new Blob(
                [
                    oneLine`
                    const publicPath = ${JSON.stringify(publicPath)};
                    globalThis = new Proxy(
                        globalThis,
                        {
                            get: (target, prop) =>
                                prop === 'location' ? publicPath : target[prop]
                        }
                    );
                    importScripts(${JSON.stringify(url.toString())});
                `,
                ],
                {
                    type: 'application/javascript',
                },
            ),
        );

        this._worker = new Worker(objectURL);
        URL.revokeObjectURL(objectURL);
    }

    get worker(): Worker {
        return this._worker;
    }
}
```

You also have to unset the `output.publicPath` in your webpack configuration. Basically what this leverages are three webpack peculiarities:
- webpack doesn't reference globals directly, it does so via `__webpack_require__.g`, which is initialized as follows:
```javascript
/******/ 	/* webpack/runtime/global */
/******/ 	(() => {
/******/ 		__webpack_require__.g = (function() {
/******/ 			if (typeof globalThis === 'object') return globalThis;
/******/ 			try {
/******/ 				return this || new Function('return this')();
/******/ 			} catch (e) {
/******/ 				if (typeof window === 'object') return window;
/******/ 			}
/******/ 		})();
/******/ 	})();
```
- auto public path resolution procedure (when `output.publicPath` is either unset or set to `auto`) is as follows:
```javascript
/******/ 	/* webpack/runtime/publicPath */
/******/ 	(() => {
/******/ 		var scriptUrl;
/******/ 		if (__webpack_require__.g.importScripts) scriptUrl = __webpack_require__.g.location + "";
/******/ 		var document = __webpack_require__.g.document;
/******/ 		if (!scriptUrl && document) {
/******/ 			if (document.currentScript)
/******/ 				scriptUrl = document.currentScript.src;
/******/ 			if (!scriptUrl) {
/******/ 				var scripts = document.getElementsByTagName("script");
/******/ 				if(scripts.length) {
/******/ 					var i = scripts.length - 1;
/******/ 					while (i > -1 && !scriptUrl) scriptUrl = scripts[i--].src;
/******/ 				}
/******/ 			}
/******/ 		}
/******/ 		// When supporting browsers where an automatic publicPath is not supported you must specify an output.publicPath manually via configuration
/******/ 		// or pass an empty string ("") and set the __webpack_public_path__ variable from your code to use your own logic.
/******/ 		if (!scriptUrl) throw new Error("Automatic publicPath is not supported in this browser");
/******/ 		scriptUrl = scriptUrl.replace(/#.*$/, "").replace(/\?.*$/, "").replace(/\/[^\/]+$/, "/");
/******/ 		__webpack_require__.p = scriptUrl;
/******/ 	})();
```
- in my case, I couldn't just overwrite the public path on the fly using `__webpack_public_path__` (which in reality is just a macro for `__webpack_require__.p`), because the chunk is not an entry and this code is getting called before any of my modules can do their thing:
```javascript
/******/ 	/* webpack/runtime/startup chunk dependencies */
/******/ 	(() => {
/******/ 		var next = __webpack_require__.x;
/******/ 		__webpack_require__.x = () => {
/******/ 			return Promise.all(["vendors-some_vendor_chunk1_redacted_for_privacy","vendors-some_vendor_chunk2_redacted_for_privacy","packages_commons_src_collections_index_ts-packages_editor_src_index_ts"].map(__webpack_require__.e, __webpack_require__)).then(next);
/******/ 		};
/******/ 	})();
```

Here's some facts about what *all 'at* means in a bigger picture:
- Everything `webpack/runtime` related is being run first. This stuff sets up `require__.g`, `require__.p` (if you set `output.publicPath` that would be just a constant string assignment, so we can't really do anything about it) and loads all chunk dependencies using, in webpack's opinion, properly set up `__webpack_require__`. All of this happens before any of our modules get even a glimpse of a chance of being run, so we can't really interfere and change `__require.p` on the fly.
- Automatic `publicPath` detection, however, determines the `require__.p` based on `require__.g.location`. Which is a global scope `location` variable/property.
- `require__.g` is resolved prioritizing `globalThis` over everything else. Granted that my company's product is designed to be run in somewhat modern browser environments, we could safely shim `globalThis.location` using a Proxy. Otherwise I think it's definitely possible to shim `self: DedicatedWorkerGlobalScope`, but it needs some testing. Proxies are a must-have here, though.

This is definitely a dirty hack, and you should probably invest in some sort of a webpack plugin that emits its own runtime module which would be a silver bullet for all of this (I think you could resolve public path using something like `self.__worker_public_path__` in that runtime module if that property is present, otherwise you could delegate that to another runtime module that is supposed to do the job. That way you'd be able to set `self.__worker_public_path__` in the blob worker).

**Reply by gigadie** [reply]
I solved it using fetch instead in order to have this to work and not to throw errors related to `importScripts` not available on `WorkerGlobalScope` when dealing with Modules.:
```typescript
export class CorsWorker {
    private readonly url: string | URL;
    private readonly options?: WorkerOptions;

    // Have this only to trick Webpack into properly transpiling
    // the url address from the component which uses it
    constructor(url: string | URL, options?: WorkerOptions) {
        this.url = url;
        this.options = options;
    }

    async createWorker(): Promise<Worker> {
        const f = await fetch(this.url);
        const t = await f.text();
        const b = new Blob([t], {
            type: 'application/javascript',
        });
        const url = URL.createObjectURL(b);
        const worker = new Worker(url, this.options);
        return worker;
    }
}
```
and then
```typescript
import { CorsWorker as Worker } from './cors-worker';

// aliasing CorsWorker to Worker makes it statically analyzable

const corsWorker = await (new Worker(
    new URL('../workers/my-worker.worker', import.meta.url),
    { type: 'module', name: 'my-worker' },
)).createWorker();
```

**Reply by CraigglesO** [reply]
This works on the newest nextJS? I hadn't upgraded in forever because I've been using the first solution.

**Reply by gigadie** [reply]
I'm using this in Angular 16, but this is purely Webpack + typescript, so I can't think of a reason why this shouldn't.

**Reply by OmarCastro** [reply]
Hello, I do not know if anyone has the following error:

So, some time passed (a year and half) after my fist answer, and I got a new error in importScripts. After analysis, I found that the URL is now prefixed with "blob:" and the relative path is absolute (e.g. `"blob:https://example.com/asset-path/file.js"` instead of `"asset-path/file.js"`). Why does it happen I do not know, the original code has not changed, only webpack was updated.

To handle those changes I had to update the code I answered previously, with comments to explain why those parts are there, here is the code:

```javascript
/**
 * A Worker that can be run on a different origin by respecting CORS
 * By default, a Worker constructor that uses an url as its parameters ignores CORS
 * and fails on a different origin.
 * This way we can run web worker on scripts served by a CDN
 */
export class CorsWorker {
  /**
   * @param {string | URL} url - worker script URL
   * @param {WorkerOptions} [options] - worker options
   */
  constructor(url, options) {
    const absoluteUrl = new URL(url, window.location.href).toString()
    const workerSource = `\
  const urlString = ${JSON.stringify(absoluteUrl)}
  const originURL = new URL(urlString)
  const isValidUrl = (urlString) => { 
    try { return Boolean(new URL(urlString, originURL)) } catch(e){ return false }  
  }
  const originalImportScripts = self.importScripts
  self.importScripts = (url) => {
    /* see note 1 below */
    if(url.startsWith("blob:") && isValidUrl(url.replace("blob:", ""))){
      const urlWithoutBlob = url.replace("blob:", "")
      const { pathname } = new URL(urlWithoutBlob, originURL)
      url = pathname && pathname.substring(1) /* see note 2 below */
    }
    originalImportScripts.call(self, new URL(url, originURL).toString())
  }
  importScripts(urlString);
`
    const blob = new Blob([workerSource], { type: 'application/javascript' })
    const objectURL = URL.createObjectURL(blob)
    this.worker = new Worker(objectURL, options)
    URL.revokeObjectURL(objectURL)
  }

  getWorker() {
    return this.worker
  }

  /*
      The notes here are outside workerSource to not increase the ObjectURL size with
    long explanations as comments

    Note 1
    ======
      Sometimes Webpack will try to import url with "blob:" prefixed and with relative
    path as absolute path.

      Not only this will cause a security error due to different url, such as a "blob:" 
    protocol (meaning a CORS error), the path will be wrong. Any of those 2 problems 
    fails the importScripts. Since we can import JS with blobs, lets just remove "blob:"
    prefix and fix the URL its content is a valid URL

    Note 2
    ======
       URL#pathname always starts with "/", we want to remove the / to be a relative path
     to the Worker file URL

   */
}

```
Hope it helps if a similar problem appears.

**Reply by mslmsl66** [reply]
I got the bellow error: 
`Uncaught DOMException: Failed to execute 'importScripts' on 'WorkerGlobalScope': The URL '/vendors-node_modules_xxx_js-node_modu-527285.async.js' is invalid.`

Your polyfill really help me! thanks a lot @OmarCastro

**Comment by Hyodori04** [comment]
I thinkt above solution not works.

**Reply by kumardeo** [reply]
Try this library: https://www.npmjs.com/package/@deox/cors-worker

**Reply by CraigglesO** [reply]
You don't need a library it's a built into webpack now:
```
module.exports = {
...
output: {
publicPath: `https://example.com/v${version}/`,
workerPublicPath: `https://example.com/v${version}/`,
}
}
```

**Reply by Manitej66** [reply]
@CraigglesO could you share the steps in detail to get webworkers working with webpack 5? I'm using nextjs and getting errors

**Reply by CraigglesO** [reply]
@Manitej66 - would it help to just see my config?
[This is my project I use it](https://github.com/Open-S2/s2maps-gpu/blob/master/configS2/webpack.config.js)
[I run it via package.json like so](https://github.com/Open-S2/s2maps-gpu/blob/5279bf8dfb56994685a3b09dec615a4dfec37c39/package.json#L34)

NOTE: I am using webpack `5.91.0` and I believe `workerPublicPath` was added around 5.81.0 or close to that.

My script uploads locally to my folder in `NuxtJS` where the main file along with all the workers are all at the same folder level.

I believe its important that you flatten the `dist` directory. I mean if you set your `workerPublicPath` to say `https://example.com/project/` but your workers are in `https://example.com/project/subFolder/` I believe this fails.

**Reply by kumardeo** [reply]
> You don't need a library it's a built into webpack now:
> 
> ```
> module.exports = {
> ...
> output: {
> publicPath: `https://example.com/v${version}/`,
> workerPublicPath: `https://example.com/v${version}/`,
> }
> }
> ```

Yeps, I knew that but it was added in recent release.

**Comment by alexander-akait** [comment]
For ECMA modules just use
```js
new Blob([`import(${JSON.stringify(url.toString())});`], { type: 'module' })
```			

in this solution https://github.com/webpack/webpack/discussions/14648#discussioncomment-1589272
